### ISSUE TITLE
Authorization objects and wildcard domains

### ISSUE THREAD

**Issue by devo-wm** [jaredhendrickson13/simple_acme_dns#37]
https://community.letsencrypt.org/t/acme-v2-production-environment-wildcards/55578
> Orders that contain both a base domain and its wildcard equivalent (e.g. *.example.com and example.com) are valid. In that case, there will be two authorization objects in the order for “[example.com 507](http://example.com/)”, one of which represents the wildcard validation and one of which represents the base domain validation.

I noticed we're stripping the `*.` before any domains that include them, for the purposes of pre-validation.  

Wondering 1) does this mean we're not making the order for the wildcard domain and 2) if not, will have an impact on the SAN field in the requested certificate?

Wondering if we just need to pre-pend `*.` after running validation. 

Happy to issue a PR if you think this is a good idea!

**Comment by jaredhendrickson13** [comment]
Hey!

This is a good find. I've verified this is a bug, but it's a little more complicated.

**TL;DR:** 
`*.` isn't actually being stripped from `domains` during validation, `domains` are being entirely rewritten with an undesired value returned by Let's Encrypt elsewhere in the code. The first attempt to request a wildcard certificate will work, but subsequent requests will not unless `domains` are reset manually and a new CSR is generated.

**Specifics:**
The `*.` _appears_ to be stripped because the entire `domains` attribute is currently being rewritten here:

https://github.com/jaredhendrickson13/simple_acme_dns/blob/9a6f809414b817c3cdf66ba52ceabf44c87af07b/simple_acme_dns/__init__.py#L560-L571

This rewrites `domains` with the domain names Let's Encrypt will check for verification tokens, not the domains included in the CSR. Let's Encrypt uses the root domain to verify wildcard certificates which is why `*.` is not present afterward. The first time a wildcard certificate is issued it will work as expected; but only because the CSR is generated before `domains` gets overwritten. After that, `domains` will need to be manually reset and a new CSR will need to be generated to receive subsequent wildcard certificates. I don't remember if there was a specific reason `domains` are overwritten here. But I'm assuming what will need to be done is remove the code that overwrites `domains`, and ensure the rest of the codebase tolerates the `*.` being present in `domains`.

A PR is definitely welcome. Otherwise, I will try to get to this as soon as I can. 

Thanks!

**Comment by devo-wm** [comment]
Thanks so much for your reply!

So in my local testing I also noticed that letsencrypt seemed to be **lstrip**-ing the `*.` in the returned`authz.body.identifier.value` as well, however, when I removed lines 623-628 

https://github.com/jaredhendrickson13/simple_acme_dns/blob/master/simple_acme_dns/__init__.py#L623


...thereby leaving the `*.` in place **before** the order was made, the resultant certificate still included the wildcard domain in the SAN field.

Can't stress how easy this package made it for me to write my script.
